File: bindings/nodejs/ffi.js
'use strict';

// koffi bindings — mirrors include/ztok.h. Every exported function is
// declared here with full C signatures so calls fail loudly on signature
// drift. _Out_ markers tell koffi that JS will read the parameter back as
// a write-output (length-1 array or struct ref).
//
// Status / enum / format codes are mirrored directly from the header so
// callers can use them as named constants instead of magic numbers.

const koffi = require('koffi');
const { loadLibztok } = require('./lib');

// --- status codes (mirror enum ztok_status) ---
const ZTOK_OK = 0;
const ZTOK_ERR_OUT_OF_MEMORY = 1;
const ZTOK_ERR_INVALID_INPUT = 2;
const ZTOK_ERR_BUFFER_TOO_SMALL = 3;
const ZTOK_ERR_INTERNAL = 99;

// --- enum kinds (mirror ztok_*_kind enums) ---
const NORMALIZER_IDENTITY = 0;
const NORMALIZER_NFC = 1;
const NORMALIZER_NFD = 2;
const NORMALIZER_NFKC = 3;
const NORMALIZER_NFKD = 4;
const NORMALIZER_BYTE_LEVEL = 5;

const PRETOK_IDENTITY = 0;
const PRETOK_CL100K = 1;
const PRETOK_TEKKEN = 2;

const MODEL_BYTE_ID = 0;

const DECODER_CONCAT = 0;
const DECODER_WORDPIECE = 1;
const DECODER_BYTE_LEVEL = 2;

// --- auto-detect format codes (mirror enum ztok_format) ---
const FORMAT_UNKNOWN = 0;
const FORMAT_TIKTOKEN = 1;
const FORMAT_HF_JSON = 2;
const FORMAT_SP_MODEL = 3;
const FORMAT_ZTM = 4;
const FORMAT_TEKKEN = 5;
const FORMAT_RWKV = 6;

// --- chunk boundary modes (mirror enum ztok_chunk_boundary) ---
const CHUNK_BOUNDARY_TOKEN = 0;
const CHUNK_BOUNDARY_CODEPOINT = 1;
const CHUNK_BOUNDARY_WORD = 2;
const CHUNK_BOUNDARY_WORD_DICT = 3;
const CHUNK_BOUNDARY_SENTENCE = 4;
const CHUNK_BOUNDARY_PARAGRAPH = 5;

// --- overlay channel kinds (mirror enum ztok_overlay_kind) ---
const OVERLAY_BYTE_START = 0;
const OVERLAY_BYTE_END = 1;
const OVERLAY_BOUNDARY = 2;
const OVERLAY_OPCODE = 3;
const OVERLAY_OPERAND = 4;
const OVERLAY_SYMBOL_REF = 5;
const OVERLAY_HUNK = 6;
const OVERLAY_PROVENANCE = 7;
const OVERLAY_USER_BASE = 0x8000;

const FORMAT_CODE_TO_NAME = {
    [FORMAT_UNKNOWN]: 'unknown',
    [FORMAT_TIKTOKEN]: 'tiktoken',
    [FORMAT_HF_JSON]: 'hf_json',
    [FORMAT_SP_MODEL]: 'sentencepiece',
    [FORMAT_ZTM]: 'ztm',
    [FORMAT_TEKKEN]: 'tekken',
    [FORMAT_RWKV]: 'rwkv',
};

// Module-level singleton — koffi caches the dlopen handle, and binding
// the same library twice is wasteful. The first import wires everything.
let cached = null;

function getLib() {
    if (cached) return cached;

    const lib = loadLibztok(koffi);

    // Struct mirroring `ztok_pipeline_config`. All four fields are the
    // numeric enum values.
    const ZtokPipelineConfig = koffi.struct('ztok_pipeline_config', {
        normalizer: 'uint32_t',
        pre_tokenizer: 'uint32_t',
        model: 'uint32_t',
        decoder: 'uint32_t',
    });

    // --- lifecycle ---
    const ztok_pipeline_new = lib.func(
        'void* ztok_pipeline_new(ztok_pipeline_config* cfg, _Out_ int* status)'
    );
    const ztok_pipeline_free = lib.func('void ztok_pipeline_free(void* p)');

    const ztok_pipeline_new_bpe_from_tiktoken = lib.func(
        'void* ztok_pipeline_new_bpe_from_tiktoken(const char* path, ztok_pipeline_config* cfg, _Out_ int* status)'
    );
    const ztok_pipeline_new_bpe_from_hf_json = lib.func(
        'void* ztok_pipeline_new_bpe_from_hf_json(const char* path, ztok_pipeline_config* cfg, _Out_ int* status)'
    );
    const ztok_pipeline_new_wordpiece_from_hf_json = lib.func(
        'void* ztok_pipeline_new_wordpiece_from_hf_json(const char* path, uint32_t unk_id, ztok_pipeline_config* cfg, _Out_ int* status)'
    );
    const ztok_pipeline_new_unigram_from_sp_model = lib.func(
        'void* ztok_pipeline_new_unigram_from_sp_model(const char* path, uint32_t unk_id, ztok_pipeline_config* cfg, _Out_ int* status)'
    );
    const ztok_pipeline_new_monster_from_file = lib.func(
        'void* ztok_pipeline_new_monster_from_file(const char* path, ztok_pipeline_config* cfg, _Out_ int* status)'
    );
    const ztok_pipeline_new_rwkv_from_file = lib.func(
        'void* ztok_pipeline_new_rwkv_from_file(const char* path, ztok_pipeline_config* cfg, _Out_ int* status)'
    );
    const ztok_pipeline_new_tekken_from_file = lib.func(
        'void* ztok_pipeline_new_tekken_from_file(const char* path, ztok_pipeline_config* cfg, _Out_ int* status)'
    );

    // --- encode / decode ---
    const ztok_encode = lib.func(
        'int ztok_encode(void* p, const char* input, size_t input_len, _Out_ uint32_t* out, size_t out_cap, _Out_ size_t* out_len)'
    );
    const ztok_decode = lib.func(
        'int ztok_decode(void* p, uint32_t* ids, size_t ids_len, _Out_ uint8_t* out, size_t out_cap, _Out_ size_t* out_len)'
    );

    // --- encode with overlay channels ---
    // Struct mirroring `ztok_overlay_channel`. `out` is a caller-owned
    // uint32_t buffer of `out_cap` entries that the encoder fills with one
    // value per token. We pass it as a void* so JS can hand koffi either
    // NULL (sizing pass) or a typed-array buffer (fill pass).
    const ZtokOverlayChannel = koffi.struct('ztok_overlay_channel', {
        kind: 'uint32_t',
        out: 'void*',
        out_cap: 'size_t',
    });
    const ztok_encode_with_overlays = lib.func(
        'int ztok_encode_with_overlays(void* p, const char* input, size_t input_len, _Out_ uint32_t* out_ids, size_t out_ids_cap, ztok_overlay_channel* channels, size_t n_channels, _Out_ size_t* out_len)'
    );

    // --- batch ---
    const ztok_encode_batch = lib.func(
        'int ztok_encode_batch(void* p, const char** inputs, size_t* input_lens, size_t n, _Out_ void** out_ids, _Out_ size_t* out_lens, uint32_t n_workers)'
    );
    const ztok_batch_pool_new = lib.func(
        'void* ztok_batch_pool_new(uint32_t n_workers, _Out_ int* status)'
    );
    const ztok_batch_pool_free = lib.func('void ztok_batch_pool_free(void* pool)');
    const ztok_batch_pool_worker_count = lib.func(
        'size_t ztok_batch_pool_worker_count(void* pool)'
    );
    const ztok_encode_batch_pooled = lib.func(
        'int ztok_encode_batch_pooled(void* p, void* pool, const char** inputs, size_t* input_lens, size_t n, _Out_ void** out_ids, _Out_ size_t* out_lens)'
    );

    const ztok_ids_free = lib.func('void ztok_ids_free(void* ids)');

    // --- Engram n-gram hashing ---
    // Raw token ids in, row-major [position][head] uint64 hashes out. The
    // single-doc path fills a caller-owned uint64 buffer; the batch path
    // hands back one ztok-allocated buffer per doc (free with
    // ztok_u64s_free). We pass id arrays/hash buffers as void* so JS can
    // marshal typed arrays or NULL.
    const ztok_ngram_hash = lib.func(
        'int ztok_ngram_hash(uint32_t* ids, size_t n_ids, uint32_t n, uint32_t heads, _Out_ uint64_t* out, size_t out_cap, _Out_ size_t* out_len)'
    );
    const ztok_ngram_hash_batch = lib.func(
        'int ztok_ngram_hash_batch(void* pool, const uint32_t** id_arrays, size_t* id_lens, size_t n_docs, uint32_t n, uint32_t heads, _Out_ void** out_hashes, _Out_ size_t* out_lens)'
    );
    const ztok_u64s_free = lib.func('void ztok_u64s_free(void* hashes)');

    // --- chunking ---
    // Struct mirroring `ztok_chunk_rec`. `ids` is a ztok-allocated buffer
    // of `ids_len` token ids (NULL when ids_len==0), released via
    // ztok_chunks_free; byte_*/token_* are half-open ranges in the
    // original input.
    const ZtokChunkRec = koffi.struct('ztok_chunk_rec', {
        ids: 'void*',
        ids_len: 'size_t',
        byte_start: 'uint32_t',
        byte_end: 'uint32_t',
        token_start: 'uint32_t',
        token_end: 'uint32_t',
    });
    const ztok_chunk = lib.func(
        'int ztok_chunk(void* pipeline, const char* text, size_t text_len, uint32_t max_tokens, uint32_t overlap, uint32_t boundary, _Out_ ztok_chunk_rec* out_chunks, size_t out_cap, _Out_ size_t* out_len)'
    );
    const ztok_chunks_free = lib.func('void ztok_chunks_free(ztok_chunk_rec* chunks, size_t n)');

    // --- auto-detect ---
    const ztok_auto_detect = lib.func('uint32_t ztok_auto_detect(const char* path)');

    // --- streaming ---
    const ztok_stream_new = lib.func(
        'void* ztok_stream_new(void* p, _Out_ int* status)'
    );
    const ztok_stream_free = lib.func('void ztok_stream_free(void* s)');
    const ztok_stream_feed = lib.func(
        'int ztok_stream_feed(void* s, const char* bytes, size_t n_bytes, _Out_ void** out_ids, _Out_ size_t* out_n_ids)'
    );
    const ztok_stream_finish = lib.func(
        'int ztok_stream_finish(void* s, _Out_ void** out_ids, _Out_ size_t* out_n_ids)'
    );

    // --- version ---
    const ztok_version = lib.func('const char* ztok_version()');

    cached = {
        koffi,
        ZtokPipelineConfig,
        ZtokOverlayChannel,
        ZtokChunkRec,
        ztok_encode_with_overlays,
        ztok_pipeline_new,
        ztok_pipeline_free,
        ztok_pipeline_new_bpe_from_tiktoken,
        ztok_pipeline_new_bpe_from_hf_json,
        ztok_pipeline_new_wordpiece_from_hf_json,
        ztok_pipeline_new_unigram_from_sp_model,
        ztok_pipeline_new_monster_from_file,
        ztok_pipeline_new_rwkv_from_file,
        ztok_pipeline_new_tekken_from_file,
        ztok_ngram_hash,
        ztok_ngram_hash_batch,
        ztok_u64s_free,
        ztok_chunk,
        ztok_chunks_free,
        ztok_encode,
        ztok_decode,
        ztok_encode_batch,
        ztok_batch_pool_new,
        ztok_batch_pool_free,
        ztok_batch_pool_worker_count,
        ztok_encode_batch_pooled,
        ztok_ids_free,
        ztok_auto_detect,
        ztok_stream_new,
        ztok_stream_free,
        ztok_stream_feed,
        ztok_stream_finish,
        ztok_version,
    };
    return cached;
}

function detectFormat(libPath) {
    const lib = getLib();
    const code = lib.ztok_auto_detect(libPath);
    return FORMAT_CODE_TO_NAME[code] || 'unknown';
}

module.exports = {
    getLib,
    detectFormat,

    // Status codes
    ZTOK_OK,
    ZTOK_ERR_OUT_OF_MEMORY,
    ZTOK_ERR_INVALID_INPUT,
    ZTOK_ERR_BUFFER_TOO_SMALL,
    ZTOK_ERR_INTERNAL,

    // Enum kinds
    NORMALIZER_IDENTITY,
    NORMALIZER_NFC,
    NORMALIZER_NFD,
    NORMALIZER_NFKC,
    NORMALIZER_NFKD,
    NORMALIZER_BYTE_LEVEL,
    PRETOK_IDENTITY,
    PRETOK_CL100K,
    PRETOK_TEKKEN,
    MODEL_BYTE_ID,
    DECODER_CONCAT,
    DECODER_WORDPIECE,
    DECODER_BYTE_LEVEL,

    // Format codes
    FORMAT_UNKNOWN,
    FORMAT_TIKTOKEN,
    FORMAT_HF_JSON,
    FORMAT_SP_MODEL,
    FORMAT_ZTM,
    FORMAT_TEKKEN,
    FORMAT_RWKV,
    FORMAT_CODE_TO_NAME,

    // Chunk boundary modes
    CHUNK_BOUNDARY_TOKEN,
    CHUNK_BOUNDARY_CODEPOINT,
    CHUNK_BOUNDARY_WORD,
    CHUNK_BOUNDARY_WORD_DICT,
    CHUNK_BOUNDARY_SENTENCE,
    CHUNK_BOUNDARY_PARAGRAPH,

    // Overlay channel kinds
    OVERLAY_BYTE_START,
    OVERLAY_BYTE_END,
    OVERLAY_BOUNDARY,
    OVERLAY_OPCODE,
    OVERLAY_OPERAND,
    OVERLAY_SYMBOL_REF,
    OVERLAY_HUNK,
    OVERLAY_PROVENANCE,
    OVERLAY_USER_BASE,
};

File: bindings/ruby/lib/ztok/ffi.rb
# frozen_string_literal: true

require "ffi"

require_relative "errors"
require_relative "lib"

module Ztok
  # Mirrors include/ztok.h via the `ffi` gem. Every C function we touch
  # is declared with full `attach_function` signatures so calls fail
  # loudly on signature drift.
  #
  # Type map (FFI -> C):
  #   :uint32           -> ztok_token_id (uint32_t)
  #   :int              -> ztok_status   (enum, C int width)
  #   :pointer          -> ztok_pipeline* / ztok_batch_pool* / ztok_stream* (opaque)
  #   :string           -> const char*   (NUL-terminated)
  #   :pointer (bytes)  -> const char*   (when carrying possibly-NUL bytes)
  #   :size_t           -> size_t
  #
  # The id-buffer header lifecycle (CRITICAL):
  #   ztok_encode_batch_pooled and ztok_stream_{feed,finish} return each
  #   id buffer as a ztok_token_id* whose bytes carry an opaque length
  #   header (see src/c_api.zig::allocIdBuf). The ONLY safe free is
  #   ztok_ids_free. NEVER call ::FFI::MemoryPointer.free on these or
  #   pass them to libc free.
  module FFI
    extend ::FFI::Library

    # Enum values (mirror ztok_*_kind in ztok.h).
    NORMALIZER_IDENTITY   = 0
    NORMALIZER_NFC        = 1
    NORMALIZER_NFD        = 2
    NORMALIZER_NFKC       = 3
    NORMALIZER_NFKD       = 4
    NORMALIZER_BYTE_LEVEL = 5

    PRETOK_IDENTITY = 0
    PRETOK_CL100K   = 1
    PRETOK_TEKKEN   = 2

    MODEL_BYTE_ID = 0

    DECODER_CONCAT     = 0
    DECODER_WORDPIECE  = 1
    DECODER_BYTE_LEVEL = 2

    # Auto-detect format codes (mirror enum ztok_format in ztok.h).
    FORMAT_UNKNOWN  = 0
    FORMAT_TIKTOKEN = 1
    FORMAT_HF_JSON  = 2
    FORMAT_SP_MODEL = 3
    FORMAT_ZTM      = 4
    FORMAT_TEKKEN   = 5
    FORMAT_RWKV     = 6

    FORMAT_CODE_TO_NAME = {
      FORMAT_UNKNOWN  => :unknown,
      FORMAT_TIKTOKEN => :tiktoken,
      FORMAT_HF_JSON  => :hf_json,
      FORMAT_SP_MODEL => :sentencepiece,
      FORMAT_ZTM      => :ztm,
      FORMAT_TEKKEN   => :tekken,
      FORMAT_RWKV     => :rwkv,
    }.freeze

    # Chunk boundary modes (mirror ztok_chunk_boundary in ztok.h).
    CHUNK_BOUNDARY_TOKEN     = 0
    CHUNK_BOUNDARY_CODEPOINT = 1
    CHUNK_BOUNDARY_WORD      = 2
    CHUNK_BOUNDARY_WORD_DICT = 3
    CHUNK_BOUNDARY_SENTENCE  = 4
    CHUNK_BOUNDARY_PARAGRAPH = 5

    # Symbolic boundary names accepted by Pipeline#chunk's `boundary:` arg.
    CHUNK_BOUNDARY_NAME_TO_CODE = {
      token:     CHUNK_BOUNDARY_TOKEN,
      codepoint: CHUNK_BOUNDARY_CODEPOINT,
      word:      CHUNK_BOUNDARY_WORD,
      word_dict: CHUNK_BOUNDARY_WORD_DICT,
      sentence:  CHUNK_BOUNDARY_SENTENCE,
      paragraph: CHUNK_BOUNDARY_PARAGRAPH,
    }.freeze

    # Overlay channel kinds (mirror ztok_overlay_kind in ztok.h). Cheap
    # channels (BYTE_START/BYTE_END/BOUNDARY/PROVENANCE) carry
    # encoder-derived values; domain channels (OPCODE/OPERAND/SYMBOL_REF/
    # HUNK) come back zero-filled until a domain plugin populates them.
    OVERLAY_BYTE_START = 0
    OVERLAY_BYTE_END   = 1
    OVERLAY_BOUNDARY   = 2
    OVERLAY_OPCODE     = 3
    OVERLAY_OPERAND    = 4
    OVERLAY_SYMBOL_REF = 5
    OVERLAY_HUNK       = 6
    OVERLAY_PROVENANCE = 7
    OVERLAY_USER_BASE  = 0x8000

    # Struct mirroring `ztok_pipeline_config` (four packed uint32 fields).
    class PipelineConfig < ::FFI::Struct
      layout :normalizer,    :uint32,
             :pre_tokenizer, :uint32,
             :model,         :uint32,
             :decoder,       :uint32
    end

    # Struct mirroring `ztok_overlay_channel`. `out` is a caller-owned
    # uint32 buffer of `out_cap` entries, filled with one value per token.
    class OverlayChannel < ::FFI::Struct
      layout :kind,    :uint32,
             :out,     :pointer,
             :out_cap, :size_t
    end

    # Struct mirroring `ztok_chunk_rec` in ztok.h. `ids` is a
    # ztok-allocated buffer of `ids_len` token ids (NULL when ids_len ==
    # 0), released via ztok_chunks_free. `byte_*` is the half-open byte
    # range in the ORIGINAL input; `token_*` the half-open token-index
    # range in the full encoding.
    class ChunkRec < ::FFI::Struct
      layout :ids,         :pointer,
             :ids_len,     :size_t,
             :byte_start,  :uint32,
             :byte_end,    :uint32,
             :token_start, :uint32,
             :token_end,   :uint32
    end

    # Load the shared library once at require-time so signature errors
    # surface immediately. Lib.resolve raises Ztok::LibraryNotFoundError
    # with a descriptive message if the library can't be located.
    ffi_lib Ztok::Lib.resolve

    # --- lifecycle ----------------------------------------------------
    attach_function :ztok_pipeline_new,
                    [:pointer, :pointer], :pointer
    attach_function :ztok_pipeline_free, [:pointer], :void

    attach_function :ztok_pipeline_new_bpe_from_tiktoken,
                    [:string, :pointer, :pointer], :pointer
    attach_function :ztok_pipeline_new_bpe_from_hf_json,
                    [:string, :pointer, :pointer], :pointer
    attach_function :ztok_pipeline_new_wordpiece_from_hf_json,
                    [:string, :uint32, :pointer, :pointer], :pointer
    attach_function :ztok_pipeline_new_unigram_from_sp_model,
                    [:string, :uint32, :pointer, :pointer], :pointer
    attach_function :ztok_pipeline_new_monster_from_file,
                    [:string, :pointer, :pointer], :pointer
    attach_function :ztok_pipeline_new_rwkv_from_file,
                    [:string, :pointer, :pointer], :pointer
    attach_function :ztok_pipeline_new_tekken_from_file,
                    [:string, :pointer, :pointer], :pointer

    # --- encode / decode ---------------------------------------------
    attach_function :ztok_encode,
                    [:pointer, :pointer, :size_t, :pointer, :size_t, :pointer], :int
    attach_function :ztok_decode,
                    [:pointer, :pointer, :size_t, :pointer, :size_t, :pointer], :int

    # --- encode with overlay channels --------------------------------
    attach_function :ztok_encode_with_overlays,
                    [:pointer, :pointer, :size_t, :pointer, :size_t,
                     :pointer, :size_t, :pointer], :int

    # --- batch -------------------------------------------------------
    attach_function :ztok_encode_batch,
                    [:pointer, :pointer, :pointer, :size_t,
                     :pointer, :pointer, :uint32], :int

    attach_function :ztok_batch_pool_new,
                    [:uint32, :pointer], :pointer
    attach_function :ztok_batch_pool_free, [:pointer], :void
    attach_function :ztok_batch_pool_worker_count, [:pointer], :size_t

    attach_function :ztok_encode_batch_pooled,
                    [:pointer, :pointer, :pointer, :pointer, :size_t,
                     :pointer, :pointer], :int

    attach_function :ztok_ids_free, [:pointer], :void

    # --- n-gram hashing (Engram) -------------------------------------
    attach_function :ztok_ngram_hash,
                    [:pointer, :size_t, :uint32, :uint32,
                     :pointer, :size_t, :pointer], :int
    attach_function :ztok_ngram_hash_batch,
                    [:pointer, :pointer, :pointer, :size_t,
                     :uint32, :uint32, :pointer, :pointer], :int
    attach_function :ztok_u64s_free, [:pointer], :void

    # --- chunking ----------------------------------------------------
    attach_function :ztok_chunk,
                    [:pointer, :pointer, :size_t, :uint32, :uint32, :uint32,
                     :pointer, :size_t, :pointer], :int
    attach_function :ztok_chunks_free, [:pointer, :size_t], :void

    # --- fingerprint -------------------------------------------------
    # ztok_status ztok_fingerprint(ztok_pipeline*, uint8_t out_32[32]).
    # out_32 is a caller-owned 32-byte buffer ztok writes back into.
    attach_function :ztok_fingerprint, [:pointer, :pointer], :int

    # --- auto-detect -------------------------------------------------
    attach_function :ztok_auto_detect, [:string], :uint32

    # --- streaming ---------------------------------------------------
    attach_function :ztok_stream_new, [:pointer, :pointer], :pointer
    attach_function :ztok_stream_free, [:pointer], :void
    attach_function :ztok_stream_feed,
                    [:pointer, :pointer, :size_t, :pointer, :pointer], :int
    attach_function :ztok_stream_finish,
                    [:pointer, :pointer, :pointer], :int

    # --- version -----------------------------------------------------
    attach_function :ztok_version, [], :string

    # Translate a path through ztok_auto_detect and return the symbol
    # form ({:tiktoken, :hf_json, :sentencepiece, :ztm, :unknown}).
    def self.detect_format(path)
      code = ztok_auto_detect(path.to_s)
      FORMAT_CODE_TO_NAME[code] || :unknown
    end

    # Materialize a libztok-owned id buffer into a plain Ruby Array,
    # then call ztok_ids_free on it. Used by Pipeline#encode_batch and
    # the streaming path. The pointer carries a length-prefix header
    # (see src/c_api.zig::allocIdBuf) so the ONLY safe free is
    # ztok_ids_free — never wrap these in a finalizer that calls libc
    # free, and never deref them after this call returns.
    def self.materialize_and_free_ids(ptr, n)
      return [] if ptr.nil? || ptr.null? || n <= 0
      ids = ptr.read_array_of_uint32(n)
      ztok_ids_free(ptr)
      ids
    end
  end
end
